include/nbl/builtin/hlsl/tgmath.hlsl

```diff
@@ -0,0 +1,231 @@
+{
+    FloatingPoint x = clamp<FloatingPoint>(_x, -0.99999, 0.99999);
+#ifdef __HLSL_VERSION
+    FloatingPoint w = -log((1.0-x) * (1.0+x));
+#else
+    FloatingPoint w = -std::log((1.0-x) * (1.0+x));
+#endif
+    FloatingPoint p;
+    if (w<5.0)
+    {
+        w -= 2.5;
+        p = 2.81022636e-08;
+        p = 3.43273939e-07 + p*w;
+        p = -3.5233877e-06 + p*w;
+        p = -4.39150654e-06 + p*w;
+        p = 0.00021858087 + p*w;
+        p = -0.00125372503 + p*w;
+        p = -0.00417768164 + p*w;
+        p = 0.246640727 + p*w;
+        p = 1.50140941 + p*w;
+    }
+    else
+    {
+#ifdef __HLSL_VERSION
+        w = sqrt(w) - 3.0;
+#else
+        w = std::sqrt(w) - 3.0;
+#endif
+        p = -0.000200214257;
+        p = 0.000100950558 + p*w;
+        p = 0.00134934322 + p*w;
+        p = -0.00367342844 + p*w;
+        p = 0.00573950773 + p*w;
+        p = -0.0076224613 + p*w;
+        p = 0.00943887047 + p*w;
+        p = 1.00167406 + p*w;
+        p = 2.83297682 + p*w;
+    }
+    return p*x;
+}
+
+
+template<typename T>
+inline T floor(NBL_CONST_REF_ARG(T) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::floor<T>(val);
+#else
+    return glm::floor<T>(val);
+#endif
+    
+}
+
+template<typename T, typename U>
+inline T lerp(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y, NBL_CONST_REF_ARG(U) a)
+{
+    return cpp_compat_intrinsics_impl::lerp_helper<T, U>::lerp(x, y, a);
+}
+
+template<typename FloatingPoint>
+inline FloatingPoint isnan(NBL_CONST_REF_ARG(FloatingPoint) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::isNan<T>(val);
+#else
+    return std::isnan(val);
+#endif
+}
+
+template<typename FloatingPoint>
+inline FloatingPoint isinf(NBL_CONST_REF_ARG(FloatingPoint) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::isInf<T>(val);
+#else
+    return std::isinf(val);
+#endif
+}
+
+template<typename  T>
+inline T pow(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y)
+{
+#ifdef __HLSL_VERSION
+    return spirv::pow<T>(x, y);
+#else
+    return std::pow(x, y);
+#endif
+}
+
+template<typename  T>
+inline T exp(NBL_CONST_REF_ARG(T) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::exp<T>(val);
+#else
+    return std::exp(val);
+#endif
+}
+
+template<typename  T>
+inline T exp2(NBL_CONST_REF_ARG(T) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::exp2<T>(val);
+#else
+    return std::exp2(val);
+#endif
+}
+
+#define DEFINE_EXP2_SPECIALIZATION(TYPE)\
+template<>\
+inline TYPE exp2(NBL_CONST_REF_ARG(TYPE) val)\
+{\
+    return _static_cast<TYPE>(1ull << val);\
+}\
+
+DEFINE_EXP2_SPECIALIZATION(int16_t)
+DEFINE_EXP2_SPECIALIZATION(int32_t)
+DEFINE_EXP2_SPECIALIZATION(int64_t)
+DEFINE_EXP2_SPECIALIZATION(uint16_t)
+DEFINE_EXP2_SPECIALIZATION(uint32_t)
+DEFINE_EXP2_SPECIALIZATION(uint64_t)
+
+template<typename  T>
+inline T log(NBL_CONST_REF_ARG(T) val)
+{
+#ifdef __HLSL_VERSION
+    return spirv::log<T>(val);
+#else
+    return std::log(val);
+#endif
+}
+
+template<typename  T>
+inline T abs(NBL_CONST_REF_ARG(T) val)
```

Comment:
Getting an `ambiguous call` when using this function. Possible overlap with this one?
https://github.com/Devsh-Graphics-Programming/Nabla/blob/c7bdd2a3ec4a5a83c485533b2d68e7c9c5b8994b/include/nbl/builtin/hlsl/complex.hlsl#L331

Reply:
could be, to be sure you can "hide"/comment out the `complex` one

Anyway the compiler usually tells you what are the options.

Example by przemek
![image](https://github.com/user-attachments/assets/b9471bdc-7943-470a-8c08-16daae2f98cf)

Reply:
it can't be because of `abs` from `complex.hlsl`. `abs` from `complex.hlsl` only alows `complex_t` on input, and since it is a templated function, no implicit conversion will be allowed

Reply:
anyway I'll let @Przemog1 fix